### PR TITLE
Adding "Optimize the image size" section with `kernel-slim` parent image.

### PR DESCRIPTION
**Pull request by uhryniuk**
Tested and works:
- locally
- turnip1
- cloud hosted

### CODE REVIEW

**Review by gkwan-ibm — CHANGES_REQUESTED** [review]
hi @uhryniuk 

- "Place `RUN features.sh` between the `COPY` commands in the `Dockerfile`." -> "Place `RUN features.sh` command after the `COPY` command that the configuration file is copied to the '/config/` destination directory ."
- no need to change `system` to `system-slim`
- no need to change `inventory` to `inventory-slim`
- no need to run `docker images -f`
- after `docker run...`, run `docker exec -it system /opt/ol/wlp/bin/productInfo featureInfo`, will see the required features only

**Review by gkwan-ibm — CHANGES_REQUESTED** [review]
@uhryniuk 
- let check the `docker images` and update its output at the end of the "Building your Docker image" subsection, for the `SIZE` column.
- at the "Optimizing the image size" section, also do the same thing. Now, the optimized size is greater than the full
- update the `testAppFinish.sh` to verify both Dockerfile and Dockerfile-full
- update the `.github/workflows/test.yml` to use superlinter v4 (see what to change from https://github.com/OpenLiberty/guide-maven-intro/pull/144/files)
- replace `CONTRIBUTING.md` as same as https://raw.githubusercontent.com/OpenLiberty/guide-microprofile-graphql/version-update-mp5/CONTRIBUTING.md

**Review by gkwan-ibm — COMMENTED** [review]
@uhryniuk 
- the statement is too long, break it with comma or multiple statments
  > This is recommended for development but when deploying to production it’s recommended to use a parent image with the kernel-slim tag to provide a bare minimum server with the ability to add the features required by the application.
 - "...destination directory . " -> "destination directory."
 - suggest to change the command `docker exec -it system /opt/ol/wlp/bin/productInfo featureInfo` to `inventory` and the output

**Review by gkwan-ibm — CHANGES_REQUESTED** [review]
@uhryniuk 
- the '/config/' should be `/config/`
  >  "...copied to the '/config/` destination directory. " 
- "... `Podman` containers:..." - "... containers:..."
  - occur in several places 
- "... new `Podman` images..." -> " ...new container images..."
  - occur in several places

**Review by gkwan-ibm — APPROVED** [review]

File: README.adoc [hunk 1]
```diff
@@ -392,6 +392,129 @@ endif::[]
 
 You can externalize the configuration of more than just the port numbers. To learn more about Open Liberty server configuration, check out the https://openliberty.io/docs/latest/reference/config/server-configuration-overview.html[Server Configuration Overview^] docs. 
 
+== Optimizing the image size
+
+As previously mentioned, the parent image used in each `Dockerfile` contains the `full` tag which includes all of the Liberty features. This is recommended for development but when deploying to production it's recommended to use a parent image with the `kernel-slim` tag. The `kernel-slim` tag provides a bare minimum server with the ability to add the features required by the application.
```
Comment (ramkumar-k-9286) [inline README.adoc]:
```suggestion
As mentioned previously, the parent image that is used in each `Dockerfile` contains the `full` tag, which includes all of the Liberty features. This parent image with the `full` tag is recommended for development, but while deploying to production it is recommended to use a parent image with the `kernel-slim` tag. The `kernel-slim` tag provides a bare minimum server with the ability to add the features required by the application.
```

File: README.adoc [hunk 2]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+`inventory/Dockerfile`
+----
+
+inventory/Dockerfile
+[source, Text, linenums, indent=0, role="code_column"]
+----
+include::finish/inventory/Dockerfile[]
+----
+
+Replace the parent image with [hotspot=kernel-slim file=0]`icr.io/appcafe/open-liberty:kernel-slim-java11-openj9-ubi` at the top of your Dockerfile. This image contains the `kernel-slim` tag that's recommended when deploying to production.
```
Comment (ramkumar-k-9286) [inline README.adoc]:
```suggestion
Replace the parent image with [hotspot=kernel-slim file=0]`icr.io/appcafe/open-liberty:kernel-slim-java11-openj9-ubi` at the top of your `Dockerfile`. This image contains the `kernel-slim` tag that is recommended when deploying to production.
```

File: README.adoc [hunk 3]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+
+Place [hotspot=features file=0]`RUN features.sh` command after the `COPY` command that the configuration file is copied to the `/config/` destination directory. The `features.sh` script adds the Liberty features that your application is required to operate.
```
Comment (ramkumar-k-9286) [inline README.adoc]: Please rephrase this sentence. it is confusing.
Place [hotspot=features file=0]`RUN features.sh` command after the `COPY` command that the configuration file is copied to the `/config/` destination directory.

File: README.adoc [hunk 4]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+
+Next, build your new `Docker` images with the `kernel-slim` parent image:
+
+[role='command']
+```
+docker build -t system:1.0-SNAPSHOT system/.
+docker build -t inventory:1.0-SNAPSHOT inventory/.
+```
+
+Verify that the images have built by executing the following command to list all the local `Docker` images:
```
Comment (ramkumar-k-9286) [inline README.adoc]:
```suggestion
Verify that the images have been built by executing the following command to list all the local `Docker` images:
```

File: README.adoc [hunk 5]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+
+After confirming that the images have been built, run the following commands to start the `Docker` containers:
+
+[role='command']
+```
+docker run -d --name system -p 9080:9080 system:1.0-SNAPSHOT
+docker run -d --name inventory -p 9081:9081 inventory:1.0-SNAPSHOT
+```
+
+Once your `Docker` containers are running, run the following command to see the list of the required features installed by `features.sh`:
```
Comment (ramkumar-k-9286) [inline README.adoc]:
```suggestion
Once your `Docker` containers are running, run the following command to see the list of required features installed by `features.sh`:
```

File: README.adoc [hunk 6]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+jsonb-2.0
+jsonp-2.0
+mpConfig-3.0
+restfulWS-3.0
+restfulWSClient-3.0
+----
+
+// static guide instructions:
+ifndef::cloud-hosted[]
+To ensure your containers are working properly, try accessing the `system` service to show the system properties of the running JVM.
```
Comment (ramkumar-k-9286) [inline README.adoc]:
```suggestion
To ensure that your containers are working properly, try accessing the `system` service to show the system properties of the running JVM.
```

File: README.adoc [hunk 7]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+See http://localhost:9080/system/properties[http://localhost:9080/system/properties]
+
+Next, replace `[system-ip-address]` with the IP address that you obtained earlier and add your localhost system properties to the `inventory` service by visting:
```
Comment (ramkumar-k-9286) [inline README.adoc]:
```suggestion
Next, replace `[system-ip-address]` with the IP address that you obtained earlier and add your localhost system properties to the `inventory` service by visiting:
```

File: README.adoc [hunk 8]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+`\http://localhost:9081/inventory/systems/pass:c[[system-ip-address]]`
+
+Then verify the addition of your localhost system properties to your `inventory` service.
```
Comment (ramkumar-k-9286) [inline README.adoc]:
```suggestion
Then, verify the addition of your localhost system properties to your `inventory` service. 
```

File: README.adoc [hunk 9]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+Next, replace `[system-ip-address]` with the IP address that you obtained earlier and add your localhost system properties to the `inventory` service by visting:
+`\http://localhost:9081/inventory/systems/pass:c[[system-ip-address]]`
+
+Then verify the addition of your localhost system properties to your `inventory` service. 
+See http://localhost:9081/inventory/systems[http://localhost:9081/inventory/systems]
+endif::[]
+
+// cloud-hosted guide instructions:
+ifdef::cloud-hosted[]
+The **system** service which shows the system properties of the running JVM, is now available to be accessed at **http://localhost:9080/system/properties**. Run the following curl command:
```
Comment (ramkumar-k-9286) [inline README.adoc]:
```suggestion
The **system** service which shows the system properties of the running JVM is now available to be accessed at **http://localhost:9080/system/properties**. Run the following curl command:
```

File: README.adoc [hunk 10]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+curl -s http://localhost:9080/system/properties | jq
+```
+
+Next, you can add your local system properties at the **http://localhost:9081/inventory/systems/[system-ip-address]** URL by replacing **[system-ip-address]** with the IP address that you obtained in the previous section. Run the following commands:
+```bash
+SYSTEM_IP=`docker inspect -f "{{.NetworkSettings.IPAddress }}" system`
+curl -s http://localhost:9081/inventory/systems/{$SYSTEM_IP} | jq
+```
+
+Then, verify the addition of your localhost system properties to the **inventory** service at **http://localhost:9081/inventory/systems** . Run the following curl command:
```
Comment (ramkumar-k-9286) [inline README.adoc]:
```suggestion
Then, verify the addition of your localhost system properties to the **inventory** service at **http://localhost:9081/inventory/systems**. Run the following curl command:
```

File: README.adoc [hunk 11]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+[source, Text, linenums, indent=0, role="code_column"]
+----
+include::finish/inventory/Dockerfile[]
+----
+
+Replace the parent image with [hotspot=kernel-slim file=0]`icr.io/appcafe/open-liberty:kernel-slim-java11-openj9-ubi` at the top of your `Dockerfile`. This image contains the `kernel-slim` tag that is recommended when deploying to production.
+
+Place [hotspot=features file=0]`RUN features.sh` command after the `COPY` command that copies the local `/config/` directory into the `Docker` image. The `features.sh` script adds the Liberty features that your application is required to operate.
+
+Ensure that you repeat these instructions again for the `system` service.
```
Comment (ramkumar-k-9286) [inline README.adoc]:
```suggestion
Ensure that you repeat these instructions for the `system` service.
```

File: README.adoc [hunk 12]
```diff
@@ -392,6 +392,129 @@ endif::[]
...
+
+Next, build your new `Docker` images with the `kernel-slim` parent image:
+
+[role='command']
+```
+docker build -t system:1.0-SNAPSHOT system/.
+docker build -t inventory:1.0-SNAPSHOT inventory/.
+```
+
+Verify that the images have been built by executing the following command to list all the local `Docker` images:
```
Comment (ramkumar-k-9286) [inline README.adoc:445]:
```suggestion
Verify that the images have been built by executing the following command to list all the local `Docker` images:
```

**Review by ramkumar-k-9286 — APPROVED** [review]

**Review by gkwan-ibm — APPROVED** [review]

**Pull request merged**